IM_calculation/source_site_dist/src_site_dist.py

```diff
@@ -218,6 +218,66 @@ def calc_rx_ry_GC2(
     return r_x, r_y
 
 
+def calc_rx_ry_GC2_vectorised(
+    srf_points: np.ndarray,
+    plane_infos: List[Dict],
+    locations: np.ndarray,
+    origin_offsets: np.ndarray = np.asarray([0]),
+):
+    """
+    Vectorised version of the GC2 function along sites and hypocentre locations.
+    Calculates Rx and Ry distances using the cross track and along track distance calculations
+    If there are multiple fault planes the Rx, Ry values are calculated for each fault plane individually, then weighted
+    according to plane length and distance to the location
+    For one fault plane this is the same as the GC1 function
+    :param srf_points: An array with shape (n, 3) giving the lon, lat, depth location of each subfault
+    :param plane_infos: A list of srf header dictionaries, as retrieved from qcore.srf.get_headers with idx=True
+    :param locations: An array with shape (m, 2) giving the lon, lat locations of each location to get Rx, Ry values for
+    :param origin_offsets: An array with shape (o) with the along strike hypocentre locations from the centre of
+    the fault for multiple realisations. If not set then only one origin will be considered at the centre position.
+    :return: Two arrays with shape (o, m) giving the Rx, Ry values for each of the given locations for each hypocentre
+    """
+    # Separate the srf points into the different plane traces
+    pnt_counts = [plane["nstrike"] * plane["ndip"] for plane in plane_infos]
+    pnt_counts.insert(0, 0)
+    pnt_counts = np.cumsum(pnt_counts)
+    pnt_sections = [
+        srf_points[pnt_counts[i] : pnt_counts[i] + header["nstrike"]]
+        for i, header in enumerate(plane_infos)
+    ]
+
+    # Adjust the offsets
+    length = sum([plane["length"] for plane in plane_infos])
+    origin_offsets = -(length / 2 + origin_offsets)
+
+    offsets = origin_offsets
+    weights = np.zeros(len(locations))
+    r_x_values = np.zeros((len(offsets), len(locations)))
+    r_y_values = np.zeros((len(offsets), len(locations)))
+    for plane_points, plane_header in zip(pnt_sections, plane_infos):
+        r_x_p, r_y_p = calc_rx_ry_GC1(plane_points, [plane_header], locations)
+        dists = np.asarray(
+            [h_dist_f(plane_points, loc[0], loc[1]) for loc in locations]
+        )
```

Comment:
To remove the list comprehension here would require refactoring/duplicating h_dist_f to allow for m*1 by 1*n matrix operations. Off the top of my head it will currently try to treat inputs as vectors and only run pairwise operations

Reply:
There might be something you could do with newaxis to make it work though